src/caffe/common.cpp

```diff
@@ -192,6 +201,26 @@ void Caffe::DeviceQuery() {
   return;
 }
 
+bool Caffe::CheckDevice(const int device_id) {
+  bool r =  ((cudaSuccess == cudaSetDevice(device_id)) &&
+            (cudaSuccess == cudaFree(0)));
+  cudaGetLastError();
+  return r;
+}
+
+int Caffe::GrabDevice(const int start_id) {
+  int count = 0;
+  int r = -1;
+  if (cudaSuccess == cudaGetDeviceCount(&count)) {
```

Comment:
Rather than putting the whole body of this method in a conditional, I think the whole thing would read better if we failed immediately if this call breaks (`CUDA_CHECK`, or is there a reason we need an error return here?), and then we can drop the `r` temporary which is making this look more complicated than it is.